### ISSUE TITLE
AES tests failing

### ISSUE THREAD

**Issue by ben-marshall** [scarv/libscarv#5]
The AES tests currently fail when running on `riscv-xcrypto`.

```sh
$> make ARCH=riscv-xcrypto build/riscv-xcrypto/work/test_aes.out
$> cat build/riscv-xcrypto/work/test_aes.out | python2
failed test_aes: enc
m == 8ccc17a63de34c6ed3d3aef688ca03c9
k == fbc2acfe862caf62c7aac705af1e7640
c == 00960000000000000000000000000000
  != 46202f3d9bdc5f07e1d7856076a1c64b
```

The exact output might be slightly different, but it does always fail.

I think there are three possible causes for this:
- AES is configured correctly for running on `riscv-xcrypto`, but is implemented incorrectly.
- AES is incorrectly configured for running on `riscv-xcrypto`
- Spike does not correctly implement the AES instruction set extensions.

Given the AES code and configuration flags, it's beyond me to understand exactly which of these three is the problem.

**Comment by phthinh** [comment]
I think we can use 8 configuration flags: `CONF_AES_ENC_EXTERN`; `CONF_AES_ENC_FINI_EXTERN`, `CONF_AES_ENC_ITER_EXTERN`, `CONF_AES_ENC_INIT_EXTERN`, `CONF_AES_DEC_EXTERN`, `CONF_AES_DEC_FINI_EXTERN`, `CONF_AES_DEC_ITER_EXTERN`, `CONF_AES_DEC_INIT_EXTERN` to narrow where the problem occurred. 
The usages of these flags are as follows:
- if `CONF_AES_{ENC/DEC}_EXTERN` is defined, `aes_{enc/dec}` is implemented using `riscv`/`xcrypto` assembly  instructions. Otherwise, these functions are implemented in `C`.
- In case of disabling `CONF_AES_{ENC/DEC}_EXTERN` flags, `aes_{enc/dec}` functions implemented in `C` call AES round sub-functions, i.e., `aes_{enc/dec}_init`, `aes_{enc/dec}_iter`, `aes_{enc/dec}_fini` if `CONF_AES_ROUND_SPLIT` is set.
- `CONF_AES_{ENC/DEC}_{INIT/ITER/FINI}` flags are defined to chose the assembly implementation of corresponding functions. Otherwise, the functions are implemented in `C`.

We know that all `C` implementations of these functions are verified. We can check the assembly implementation of specific function on Spike by disabling all of these configuration flags except the flag for the functions that we want to check.

So far, I have checked that `CONF_AES_{ENC/DEC}_INIT` works properly. But enabling the assembly implementation of `aes_enc_fini` function, the test was failed. So, I think the problem may come from the   assembly instructions used in this function.

**Comment by ben-marshall** [comment]
Fixed by scarv/riscv-isa-sim@2d63e5ba7
- Bug in scatter/gather implementation meant offsets came from the wrong register.
